### ISSUE TITLE
code review

### ISSUE THREAD

**Issue by GaelGil** [GaelGil/notebooks#15]
@camoverride I have worked on back propagation in this notebook [here](https://github.com/GaelGil/notebooks/blob/master/back-propagation/neural_network.ipynb). I wanted to know if this looks good, if there is a mistake somewhere, or if I am missing somewhere. Mainly because I am not sure it's working correctly. Any feedback or next steps are welcome. 

One of the things I am having trouble with is: Is the h2 the output layer? Or is the output layer equal to `sigmoid(h2*w3 + b3)`. `(h2=activations,w3=weights,b3=biases)`. From what I've read it seems that the h2 is just passed to the output layer but that doesn't seem right. 

Another concern/question is: When I feed forward an example the output activations/activations are all very low. I'm not sure why or where to look for an error for this but I don't think that is a good thing because that would mean the network is not confident the input is part of any class. 

I have implemented back propagation the way it is in [this notebook](https://github.com/GaelGil/notebooks/blob/master/back-propagation/back_propagatio_math.ipynb). Let me know if there are any questions about what my code is doing or if I should add more comments.

**Comment by camoverride** [comment]
I'm not going to dig into the code just yet because I think you should do a little research first:

> From what I've read it seems that the h2 is just passed to the output layer but that doesn't seem right.

Technically, this is an engineering problem, so there is no "correct" answer -- whatever works best is best. Play around and see if you get better performance with an activation function on the last layer. You should have __training__ and __test__ sets, which means you can calculate the loss on these and compare the effectiveness of different models you make. Also, you can look to keras for inspiration.

> When I feed forward an example the output activations/activations are all very low.

Read up on [the vanishing gradient problem](https://en.wikipedia.org/wiki/Vanishing_gradient_problem) and amend your network by either 1) simplifying it to a few neurons and just a couple layers or 2) using a different activation function. The sigmoid function is the _oldest_ activation function, and is impractical for various reasons. Also, consider whether your batch size (an important hyper-parameter) is influencing learning. What is the effect of a small vs large batch size when training and how might this influence the gradients you get?

**Comment by GaelGil** [comment]
> Technically, this is an engineering problem, so there is no "correct" answer -- whatever works best is best. Play around and see if you get better performance with an activation function on the last layer. You should have **training** and **test** sets, which means you can calculate the loss on these and compare the effectiveness of different models you make. Also, you can look to keras for inspiration.

I have read more on this and I have decided what I am going to do for this part. Once I have calculated my values for h2 I will then pass that to the softmax.

**Comment by GaelGil** [comment]
> Read up on [the vanishing gradient problem](https://en.wikipedia.org/wiki/Vanishing_gradient_problem) and amend your network by either 1) simplifying it to a few neurons and just a couple layers or 2) using a different activation function. The sigmoid function is the _oldest_ activation function, and is impractical for various reasons. Also, consider whether your batch size (an important hyper-parameter) is influencing learning. What is the effect of a small vs large batch size when training and how might this influence the gradients you get?

@camoverride I have updated the network and it now works. Previously I use the sigmoid function as the activation. The network also had one hidden layer with four neurons and a second hidden layer with three neurons. What I change was that I now used the relu activation and the first hidden layer now has two neurons instead of four. I also don't use an activation from the `h2 -> output`, the output is now just `softmax(h2)`

**Comment by camoverride** [comment]
@GaelGil - if/when you present this to me, be prepared to __prove__ that your network works.

For instance, it's not enough to say "my network correctly classifies images 90% of the time" -- what if 90% of the images are class `1` and your network outputs `1` every time? Were there other metrics you used? Why or why not? When you looked at the model's errors, was there a specific pattern you noticed and how would you correct it? Did you look at a confusion matrix? Also, when dealing with NN's, it's not enough to show just one model - you also want to prove that your model is more effective than a simple logistic regression. Also, how did the model's performance change as you increased the number of training epochs? How long did it take to train? Was there overfitting? Did it require a GPU? These (and more) are all things you'll have to show me.

**Comment by GaelGil** [comment]
> @GaelGil - if/when you present this to me, be prepared to **prove** that your network works.
> 
> For instance, it's not enough to say "my network correctly classifies images 90% of the time" -- what if 90% of the images are class `1` and your network outputs `1` every time? Were there other metrics you used? Why or why not? When you looked at the model's errors, was there a specific pattern you noticed and how would you correct it? Did you look at a confusion matrix? Also, when dealing with NN's, it's not enough to show just one model - you also want to prove that your model is more effective than a simple logistic regression. Also, how did the model's performance change as you increased the number of training epochs? How long did it take to train? Was there overfitting? Did it require a GPU? These (and more) are all things you'll have to show me.

Ill work on adding visuals and ways to test the network work wells together.

**Comment by GaelGil** [comment]
I am closing this issue since the main topic, issue has been solved.
